### ISSUE TITLE
REQUEST: Make MDPrivateKeys flags clearer/order non-specific

### ISSUE THREAD

**Issue by Jon8RFC** [icing/mod_md#217]
To get both an secp384r1 and RSA4096 cert simultaneously, I wasn't aware of specific ordering of RSA being first and the fact that secp384r1 was one flag and RSA requires two separate flags.  Either that, or the fact that one cert failed and the other succeeded causing rate limiting was problematic in determining how to get an RSA cert at 4096 bits alongside an secp384r1 cert.  I believe I set secp384r1 first, initially.

The RSA flag defaults to 2048 or to server-preference of letsencrypt.  A space and 2048 also sets 2048.  I suggest requiring RSA and its encryption strength be one flag which would also prevent unexpected defaults being applied which may change while the httpd.conf configuration may not: RSA4096, RSA2048, etc, similarly to secp384r1.

In the event that even more certificates are supported or become the norm for legacy reasons or changes throughout openssl 3.0, it would be easier to understand one cert being only ever being one flag.  If the order is specific, changing that to not requiring RSA before secp384r1 would be helpful, but that may have been the rate limiting causing failures, which I don't recall.

Thanks for the great module!

**Comment by icing** [comment]
There is some backward compatibility game here. But I like the notion of allowing `rsaNNNN` as alternate names.

The module will handle `RSA NNNN` as first or second (or nth) specification. There are test cases for this. If I did not catch a particular case, please provide an example.

**Comment by Jon8RFC** [comment]
I've tried to test and see if I can get it to fail like before, but I'm getting a new problem now.  The apache mod_status page displays "ongoing" for the certs for quite some time, and then it displays this:
Error[Internal error (specific information not available)]: Unable to retrieve secp384r1 certificate chain. Next run in ~5 seconds

Is it possible that LetsEncrypt banned me because of the past failures and the repeated queries (it retries in about 3 seconds, then 5, then slightly higher and slightly higher each time, seemingly infinitely)?  I never received an email mentioning anything.

**Comment by icing** [comment]
Did you try the `md-status` handler and the JSON it produces? It will give you more detail, including some mdomain specific logging. 

The logging itself you will also find in the staging directory of the mdomain.

**Comment by Jon8RFC** [comment]
Comparing to archives (thanks for having it save those!), it looks like it skips over validation, or letsencrypt simply doesn't attempt to validate, and then attempts to grab the cert anyway.  Strangely, not every archived job.md has the line "Setting up challenge 'tls-alpn-01' for domain jon8rfc.homeip.net", but some do:

"last": {
    "status": 20014,
    "status-description": "Internal error (specific information not available)",
    "detail": "Unable to retrieve rsa certificate chain.",
    "activity": "Retrieving rsa certificate chain for jon8rfc.homeip.net"
  },
  "log": {
    "entries": [
      {
        "when": "Thu, 28 May 2020 11:10:45 GMT",
        "type": "renewal-error",
        "detail": "Unable to retrieve rsa certificate chain."
      },
      {
        "when": "Thu, 28 May 2020 11:10:45 GMT",
        "type": "progress",
        "detail": "Retrieving rsa certificate chain for jon8rfc.homeip.net: Unable to retrieve rsa certificate chain."
      },
      {
        "when": "Thu, 28 May 2020 11:10:45 GMT",
        "type": "progress",
        "detail": "Retrieving rsa certificate chain for jon8rfc.homeip.net"
      },
      {
        "when": "Thu, 28 May 2020 11:10:15 GMT",
        "type": "progress",
        "detail": "Monitoring challenge status for jon8rfc.homeip.net"
      },
      {
        "when": "Thu, 28 May 2020 11:10:15 GMT",
        "type": "progress",
        "detail": "Setting up challenge 'tls-alpn-01' for domain jon8rfc.homeip.net"
      },
      {
        "when": "Thu, 28 May 2020 11:10:15 GMT",
        "type": "progress",
        "detail": "Starting challenges for domains"
      },
      {
        "when": "Thu, 28 May 2020 11:10:15 GMT",
        "type": "progress",
        "detail": "Loaded order from staging"
      },
      {
        "when": "Thu, 28 May 2020 11:10:14 GMT",
        "type": "progress",
        "detail": "Selecting account to use for jon8rfc.homeip.net"
      },
      {
        "when": "Thu, 28 May 2020 11:10:14 GMT",
        "type": "progress",
        "detail": "Driving ACME to renew rsa certificate for jon8rfc.homeip.net"
      },
      {
        "when": "Thu, 28 May 2020 11:10:14 GMT",
        "type": "progress",
        "detail": "Contacting ACME server for jon8rfc.homeip.net at https://acme-v02.api.letsencrypt.org/directory"
      },
      {
        "when": "Thu, 28 May 2020 11:10:14 GMT",
        "type": "progress",
        "detail": "Assessing current status"
      },
      {
        "when": "Thu, 28 May 2020 11:10:14 GMT",
        "type": "progress",
        "detail": "Checking staging area"
      },
      {
        "when": "Thu, 28 May 2020 11:10:14 GMT",
        "type": "progress",
        "detail": "Retrieving rsa certificate chain for jon8rfc.homeip.net"
      },
      {
        "when": "Thu, 28 May 2020 11:10:14 GMT",
        "type": "starting"

**Comment by Jon8RFC** [comment]
It might be LetsEncrypt being down.  I checked yesterday when it first happened, but there wasn't any mention of a problem.  I've checked again, and now it's showing that it was down yesterday, AFTER I had the failure.  Maybe there's still an issue:
https://statusgator.com/services/lets-encrypt

**Comment by icing** [comment]
You could check if you have the same problems with  the staging api (that also has less restrictions on quotas). Until you get everything working.

You can also configure

```
LogLevel md:trace5
```

for _very detailed_ logging in your apache error.log to see even how request/responses from letsencrypt are.

**Comment by Jon8RFC** [comment]
Thanks for the help.  I switched to staging and the same problem seems to occur.

From what I read in a post of yours on letsencrypt, could this be related to OCSP stapling being enforced via the cert?  That's the best I could find through searches.  The only thing which I think has changed were the May Windows updates being installed on my Hyper-V host.

I tried to strip down what looks like the errors and removed tokens and file paths, which I'd rather not post publicly.  I'm happy to personally email you a complete log if it's something you have time to take a look at.  I don't know what is timing out, all of a sudden.

[Fri May 29 23:52:45.941835 2020] [md:debug] [pid 1136:tid 1288] md_acme_authz.c(215): ACME server authz: challenge 'pending' for jon8rfc.homeip.net at 
[Fri May 29 23:52:45.941835 2020] [md:debug] [pid 1136:tid 1288] md_acme_order.c(510): (11)Resource temporarily unavailable: jon8rfc.homeip.net: status pending at 
[Fri May 29 23:52:45.941835 2020] [md:debug] [pid 1136:tid 1288] md_acme_order.c(552): (70007)The timeout specified has expired: jon8rfc.homeip.net: checked authorizations
[Fri May 29 23:52:45.941835 2020] [md:debug] [pid 1136:tid 1288] md_result.c(254): md[jon8rfc.homeip.net] while[Monitoring challenge status for jon8rfc.homeip.net]mod_md.c(240): store event=0 on file  (group 4)
[Fri May 29 23:52:45.943789 2020] [md:debug] [pid 1136:tid 1288] md_acme_drive.c(756): jon8rfc.homeip.net: retrieving rsa certificate chain
[Fri May 29 23:52:45.943789 2020] [md:error] [pid 1136:tid 1288] (20014)Internal error (specific information not available): jon8rfc.homeip.net: asked to retrieve chain, but no certificate url part of order
[Fri May 29 23:52:45.943789 2020] [md:debug] [pid 1136:tid 1288] md_result.c(254): (20014)Internal error (specific information not available): md[jon8rfc.homeip.net] while[Retrieving rsa certificate chain for jon8rfc.homeip.net] detail[Unable to retrieve rsa certificate chain.]
[Fri May 29 23:52:45.943789 2020] [md:debug] [pid 1136:tid 1288] md_reg.c(1117): (20014)Internal error (specific information not available): jon8rfc.homeip.net: staging done
[Fri May 29 23:52:45.943789 2020] [md:error] [pid 1136:tid 1288] (20014)Internal error (specific information not available): AH10056: processing jon8rfc.homeip.net: Unable to retrieve rsa certificate chain.
[Fri May 29 23:52:45.943789 2020] [md:info] [pid 1136:tid 1288] AH10057: jon8rfc.homeip.net: encountered error for the 4. time, next run in 40 seconds

**Comment by Jon8RFC** [comment]
Must've been a fluke.  Nothing has changed on my end other than repetitive attempts.  It magically worked today with staging (and it was failing yesterday with staging) and I commented out 
<pre>MDCertificateAuthority https://acme-staging-v02.api.letsencrypt.org/directory</pre>
Yet it seems to have kept the staging cert which it successfully retrieved.  That's the only flag I set for staging.  Did I do something incorrectly?  My understanding was that staging was for dry/test runs and nothing actually happened in the real world.
